torch/csrc/cuda/comm.cpp

```diff
@@ -34,38 +35,75 @@ struct unique_type_checker {
   bool unique = true;
 };
 
-std::vector<Tensor> broadcast(const Tensor& tensor, IntArrayRef devices) {
-  if (tensor.is_cuda() && tensor.get_device() != devices[0])
-    throw std::runtime_error("device of broadcasted tensor must appear as the "
-                             "first on devices list");
-  std::vector<Tensor> tensors;
-  tensors.reserve(devices.size());
+// ***************** Broadcast *******************
+//
+// Broadcast a source tensor (CPU or CUDA) to a list of CUDA devices, or CUDA
+// tensors on one or more devices.
+
+// no checks
+static inline
+std::vector<Tensor>& _broadcast_out_impl(const Tensor& tensor, std::vector<Tensor> &out_tensors) {
 #ifdef USE_NCCL
-  if (nccl::is_available({tensor})) {
-    tensors.push_back(tensor);
-    for (auto device : devices.slice(1)) {
-      tensors.push_back(
-          at::empty(tensor.sizes(),
-          tensor.options().device(at::Device(kCUDA, device))));
-    }
-    nccl::broadcast(tensors);
+  std::vector<Tensor> nccl_list;
+  nccl_list.reserve(out_tensors.size() + 1);
+  nccl_list.push_back(tensor);
+  for (auto& out_tensor : out_tensors) {
+    nccl_list.push_back(out_tensor);
+  }
```

Comment:
Will it be better to move these lines into the `if` branch below? So that when nccl is not available but using `USE_NCCL=1`, we don't have to create this vector?

Reply:
:) But we need to use this `vector<Tensor>` to test if NCCL can accept them.